_posts/2018-02-26-success-with-serverless.md

```diff
@@ -0,0 +1,42 @@
+---
+layout: post
+status: publish
+published: true
+title: 'Success With Serverless'
+author:
+  display_name: Scott Davis
+  email: stdavis@utah.gov
+date: 2018-02-26 23:45:28
+categories:
+- Developer
+tags:
+- nodejs
+---
+
+In this post, I'd like to describe a recent success story that we've had using [serverless computing](https://en.wikipedia.org/wiki/Serverless_computing). This new(ish) technology has some exciting advantages over traditional solutions. We were looking for an excuse to give it a try and found a great fit.
+
+### The Problem
+[AGRC's base maps]({{ "/data/base-map-and-imagery/" | prepend: site.baseurl }}) (including the [Google imagery]({{ "/discover/license/" | prepend: site.baseurl }})) are served via a custom server application called Giza. Part of the advantage of this solution is the ability to secure and track usage via quad-words. These unique words are assigned to a specific user and are locked down to a specific domain or IP address. For example, if my quad-word is locked down to <code>mydomain.com</code>, then requests originating from any other domain or IP address are blocked by the server. This prevents unauthorized access of licensed content as well as allows AGRC to track analytics.
+
+This quad-word system works great until you try to use one of Esri's out-of-the-box print services ([here's an example of one](http://mapserv.utah.gov/arcgis/rest/services/Utilities/PrintingTools/GPServer)). When you send a web map to this service, it reconstructs all of the layers on the server. This causes requests for base maps to be sent from your ArcGIS Server box, rather than your user's browser (with your domain as the referrer). We do allow wide open quad-words (not locked down to any domain/IP) to be used by those who need to make requests from servers or other local machines. However, these wide open quad-words cannot be used in web applications due to the fact that they could be copied and used by unauthorized users.
+
+### Swing and a Miss
+Our original solution to this problem was a [custom geoprocessing service that was deployed via ArcGIS Server](https://github.com/agrc/print-proxy). Basically, this service acted as a proxy to a traditional print service switching out the secured quad-word with a wide open one allowing the traditional print service to successfully make requests.
+
+While we did make this solution work, it was not ideal. Geoprocessing services in general are a pain to work with. Debugging can be particuarly challenging. Because of the potential strain on our server, we asked our users to publish this service on their own servers. This added additional technical debt for them to deploy and maintain (assuming that they had an ArcGIS Server instance at all).
```

Comment:
_"Geoprocessing services in general are a pain to work with. Debugging can be particuarly challenging."_: Revise to "Geoprocessing services, in general, are a pain to work with, but debugging can be particularly challenging."